R/reportToREMINDenergyEfficiency.R

```diff
@@ -14,65 +13,26 @@
-  hybrids[, techWithoutHybrid := "BEV"]
-  f35_fe2es <- rbind(f35_fe2es, hybrids)
-  #prepare weight
-  #use FE demand as weight but keep hybrid electric in order to aggregate accurately
-  weight <- copy(f35_fe2es)[, c("unit", "variable") := NULL]
-  fleetESdemand <- copy(fleetESdemand)[, c("unit", "variable") := NULL]
-  setnames(fleetESdemand, "value", "ESdemand")
-  fleetESdemand[, techWithoutHybrid := technology]
-  fleetESdemand[technology == "Hybrid electric", techWithoutHybrid := "Liquids"]
-  hybrids <- fleetESdemand[technology == "Hybrid electric"]
-  hybrids[, techWithoutHybrid := "BEV"]
-  fleetESdemand <- rbind(fleetESdemand, hybrids)
-  fleetESdemand[technology == "Hybrid electric" & techWithoutHybrid == "Liquids", ESdemand := ESdemand * (1 - hybridElecShare)]
-  fleetESdemand[technology == "Hybrid electric" & techWithoutHybrid == "BEV", ESdemand := ESdemand * (hybridElecShare)]
-  weight <- merge(weight, fleetESdemand, by = intersect(names(weight), names(fleetESdemand)))
-  weight[, FEdemand := value * ESdemand][, c("value", "ESdemand") := NULL]
-  byCols <- names(weight)
-  byCols <- byCols[!byCols %in% c("value")]
 
-  f35_fe2es[, value := 1 / value][, unit := "(p|t)km/MJ"]
-  # convert to trn (pkm|tkm)/TWa
-  f35_fe2es[, value := value * 10^-12/MJtoTwa]
-  f35_fe2es <- merge(f35_fe2es, weight, by = intersect(names(f35_fe2es), names(weight)))                                  # nolint: object_name_linter
-  fe2esMap <- unique(helpers$mapEdgeToREMIND[, c("all_teEs", "univocalName", "technology")])
-  fe2esMap <- fe2esMap[!is.na(all_teEs)]
-  setnames(fe2esMap, "technology", "techWithoutHybrid")
-  f35_fe2es <- merge(f35_fe2es, fe2esMap, by = c("univocalName", "techWithoutHybrid"), all.x = TRUE)                                                         # nolint: object_name_linter
-  f35_fe2es[, sumFE := sum(FEdemand), by = c("region", "period", "all_teEs")]
-  # Remove weight if whole branch has zero demand to keep data
-  f35_fe2es[sumFE == 0, FEdemand := 1]
-  f35_fe2es[, sumFE := sum(FEdemand), by = c("region", "period", "all_teEs")]
-  f35_fe2es <- f35_fe2es[, .(value = sum(value * FEdemand / sumFE)),  by = c("region", "period", "all_teEs")]                           # nolint: object_name_linter
-  checkForNAsAndDups(f35_fe2es, "f35_fe2es", "reportToREMINDenergyEfficiency()")
-  f35_fe2es <- prepareForREMIND(f35_fe2es, demScen, SSPscen, transportPolScen)
-  setnames(f35_fe2es, c("period", "region"), c("tall", "all_regi"))
+  ES <- reportToREMINDesDemand(fleetESdemand, hybridElecShare, timeResReporting, demScen, SSPscen, transportPolScen, helpers)
+  FE <- reportToREMINDfinalEnergyDemand(fleetFEdemand, timeResReporting, demScen, SSPscen, transportPolScen, helpers)[, c("all_enty", "all_in") := NULL]
+  setnames(FE, "value", "FE")
+  setnames(ES, "value", "ES")
+  f35_fe2es <- merge(ES, FE, by = intersect(names(FE), (names(ES))))
+  # Insert small number instead of zero
+  f35_fe2es <- f35_fe2es[!ES < 1e-7]
```

Comment:
@robertpietzcker The differences in the energy efficiency values for freight are coming from a change in how the energy efficiency is caluclated. They appear solely, where the energy service demand is anyway zero. 
Previously I used the energy intensity and aggregated it to all_teEs level by using the energy service demand as weight.
Where the whole branch has zero energy service demand, a weight of 1 was used, so that the energy intensity value does not disappear (the CES tree cannot deal with an energy efficiency of zero). Now I used FE and ES on te_Es level including the active modes to calculate the energy efficiency.  To get rid of the zeros I used another method: I removed them beforehand and interpolated the missing values afterwards. Filtering out the zero demand values in advance ensures that we do not also interpolate NAs, which would indicate errors in the model if they occur.

So in summary: The changes come from a different way of dealing with zero demand values and don't affect the model results.